File: libs/ai-kernel/src/domain/rag/retriever.rs
use std::cmp::Ordering;

use crate::{
    domain::llm::cache,
    models::knowledge_base::{KnowledgeDocument, KnowledgeSearchResult},
};

pub fn search(
    query: &str,
    documents: &[KnowledgeDocument],
    top_k: usize,
    min_score: f32,
) -> Vec<KnowledgeSearchResult> {
    let query_embedding = crate::domain::rag::embedder::embed_text(query);
    search_with_embedding(&query_embedding, documents, top_k, min_score)
}

pub fn search_with_embedding(
    query_embedding: &[f32],
    documents: &[KnowledgeDocument],
    top_k: usize,
    min_score: f32,
) -> Vec<KnowledgeSearchResult> {
    let mut hits = Vec::new();

    for document in documents {
        for chunk in &document.chunks {
            let score = cache::cosine_similarity(&query_embedding, &chunk.embedding);
            if score < min_score {
                continue;
            }

            hits.push(KnowledgeSearchResult {
                knowledge_base_id: document.knowledge_base_id,
                document_id: document.id,
                document_title: document.title.clone(),
                chunk_id: chunk.id.clone(),
                score,
                excerpt: chunk.text.clone(),
                source_uri: document.source_uri.clone(),
                metadata: chunk.metadata.clone(),
            });
        }
    }

    hits.sort_by(|left, right| {
        right
            .score
            .partial_cmp(&left.score)
            .unwrap_or(Ordering::Equal)
    });
    hits.truncate(top_k.max(1));
    hits
}

File: libs/ai-kernel-go/domain/rag/retriever.go
package rag

import (
	"sort"

	"github.com/openfoundry/openfoundry-go/libs/ai-kernel-go/domain/llm"
	"github.com/openfoundry/openfoundry-go/libs/ai-kernel-go/models"
)

// Search retrieves the top-K knowledge hits matching the query.
// Mirrors Rust src/domain/rag/retriever.rs::search — embeds the
// query then delegates to SearchWithEmbedding.
func Search(query string, documents []models.KnowledgeDocument, topK uint32, minScore float32) []models.KnowledgeSearchResult {
	q := EmbedText(query)
	return SearchWithEmbedding(q, documents, topK, minScore)
}

// SearchWithEmbedding scores every chunk against the precomputed
// embedding via cosine similarity, drops everything below minScore,
// sorts by score desc, and truncates to max(topK, 1).
func SearchWithEmbedding(queryEmbedding []float32, documents []models.KnowledgeDocument, topK uint32, minScore float32) []models.KnowledgeSearchResult {
	hits := []models.KnowledgeSearchResult{}
	for _, doc := range documents {
		for _, chunk := range doc.Chunks {
			score := llm.CosineSimilarity(queryEmbedding, chunk.Embedding)
			if score < minScore {
				continue
			}
			hits = append(hits, models.KnowledgeSearchResult{
				KnowledgeBaseID: doc.KnowledgeBaseID,
				DocumentID:      doc.ID,
				DocumentTitle:   doc.Title,
				ChunkID:         chunk.ID,
				Score:           score,
				Excerpt:         chunk.Text,
				SourceURI:       doc.SourceURI,
				Metadata:        chunk.Metadata,
			})
		}
	}
	sort.SliceStable(hits, func(i, j int) bool { return hits[i].Score > hits[j].Score })
	limit := int(topK)
	if limit < 1 {
		limit = 1
	}
	if len(hits) > limit {
		hits = hits[:limit]
	}
	return hits
}
